Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -0,0 +1,428 @@
+lemma IsBalanced.append {q : DyckPath} (bq : q.IsBalanced) : (p ++ q).IsBalanced := by
+  simp only [IsBalanced, take_append_eq_append_take, count_append]
+  exact ⟨by rw [bp.1, bq.1], fun _ ↦ add_le_add (bp.2 _) (bq.2 _)⟩
+
+/-- If `x` is a balanced Dyck path then so is `(x)`. -/
+lemma IsBalanced.nest : (↑[U] ++ p ++ ↑[D]).IsBalanced := by
+  simp only [IsBalanced, take_append_eq_append_take, count_append]
+  refine' ⟨by rw [bp.1]; simp [add_comm], fun i ↦ _⟩
+  rw [← add_rotate (count D _), ← add_rotate (count U _)]
+  apply add_le_add _ (bp.2 _)
+  cases' i.eq_zero_or_pos with hi hi; · simp [hi]
+  simp_rw [take_all_of_le (show [U].length ≤ i by rwa [length_singleton]),
+    count_singleton', ite_true, ite_false]
+  rw [add_comm _ 0]
+  exact add_le_add (zero_le _) ((count_le_length _ _).trans (by simp))
+
+/-- A balanced Dyck path is split into two parts. If one part is balanced, so is the other. -/
+lemma IsBalanced.split (i : ℕ) : IsBalanced (p.take i) ↔ IsBalanced (p.drop i) := by
+  rw [IsBalanced, ← p.take_append_drop i] at bp
+  simp_rw [take_append_eq_append_take, count_append] at bp
+  obtain ⟨b1, b2⟩ := bp
+  constructor <;> intro bh
+  · refine' ⟨by rw [bh.1] at b1; omega, fun j ↦ _⟩
+    replace b2 := b2 (j + (p.take i).length)
+    rwa [take_all_of_le (by omega), bh.1, add_le_add_iff_left, Nat.add_sub_cancel] at b2
+  · refine' ⟨by rw [bh.1] at b1; omega, fun j ↦ _⟩
+    cases' le_or_gt (p.take i).length j with hj hj
+    · rw [take_all_of_le hj]; rw [bh.1] at b1; omega
+    · replace b2 := b2 j
+      simp_rw [show j - (p.take i).length = 0 by omega, take_zero, count_nil, add_zero] at b2
+      exact b2
+
+/-- If `p.take i` satisfies the equality part of `IsBalanced`, it also satisfies the other part
+and is balanced itself. -/
+lemma IsBalanced.from_take {i : ℕ} (h : (p.take i).count U = (p.take i).count D) :
+    IsBalanced (p.take i) := ⟨h, fun j ↦ by rw [take_take]; apply bp.2⟩
+
+variable (np : p ≠ [])
+
+/-- The first element of a nonempty balanced Dyck path is `U`. -/
+lemma IsBalanced.head_U : p.head np = U := by
+  cases' p with h _; · tauto
+  by_contra f; simpa [(step_cases h).resolve_left f] using bp.2 1
+
+/-- The last element of a nonempty balanced Dyck path is `D`. -/
+lemma IsBalanced.getLast_D : p.getLast np = D := by
+  by_contra f; have s := bp.1
+  rw [← dropLast_append_getLast np, (step_cases _).resolve_right f] at s
+  simp_rw [dropLast_eq_take, count_append, count_singleton', ite_true, ite_false] at s
+  have t := bp.2 p.length.pred; omega
+
+private lemma denest_aux : p.take 1 = [p.head np] ∧ p.dropLast.take 1 = [p.head np] ∧
+    p.length.pred = p.length.pred.pred + 1 := by
+  cases' p with _ q; · tauto
+  cases' q
+  · replace bp := bp.length_even; norm_num [even_iff_two_dvd] at bp
+  · refine' ⟨_, _, _⟩
+    · rw [take_cons, take_zero, head_cons]
+    · rw [dropLast_cons₂, take_cons, take_zero, head_cons]
+    · rfl
+
+lemma append_tail_dropLast_append : p = [U] ++ tail (dropLast p) ++ [D] := by
```

Comment:
```suggestion
lemma cons_tail_dropLast_concat : U :: tail (dropLast p) ++ [D] = p := by
```
?

Reply:
For consistency with how I represent adding `D` at the end I'd prefer my original formulation with `[U] ++ ...`

Reply:
But `a :: l` is simp normal form for `[a] ++ l`, while `l ++ [a]` is simp normal form for `concat l a`, so you are putting yourself in inconsistency to the rest of the library

Reply:
Done.

Reply:
(I must say I find this asymmetry a bit stupid, but I haven't tried to fix it because it would make swapping the names `append` and `concat` harder in the future)